lightning/src/ln/channelmanager.rs

```diff
@@ -7906,6 +7704,346 @@ where
+			msg: "The final CLTV expiry is too soon to handle",
+		});
+	}
+	if (!allow_underpay && onion_amt_msat > amt_msat) ||
+		(allow_underpay && onion_amt_msat >
+		 amt_msat.saturating_add(counterparty_skimmed_fee_msat.unwrap_or(0)))
+	{
+		return Err(InboundOnionErr {
+			err_code: 19,
+			err_data: amt_msat.to_be_bytes().to_vec(),
+			msg: "Upstream node sent less than we were supposed to receive in payment",
+		});
+	}
+
+	let routing = if let Some(payment_preimage) = keysend_preimage {
+		// We need to check that the sender knows the keysend preimage before processing this
+		// payment further. Otherwise, an intermediary routing hop forwarding non-keysend-HTLC X
+		// could discover the final destination of X, by probing the adjacent nodes on the route
+		// with a keysend payment of identical payment hash to X and observing the processing
+		// time discrepancies due to a hash collision with X.
+		let hashed_preimage = PaymentHash(Sha256::hash(&payment_preimage.0).into_inner());
+		if hashed_preimage != payment_hash {
+			return Err(InboundOnionErr {
+				err_code: 0x4000|22,
+				err_data: Vec::new(),
+				msg: "Payment preimage didn't match payment hash",
+			});
+		}
+		if !accept_mpp_keysend && payment_data.is_some() {
+			return Err(InboundOnionErr {
+				err_code: 0x4000|22,
+				err_data: Vec::new(),
+				msg: "We don't support MPP keysend payments",
+			});
+		}
+		PendingHTLCRouting::ReceiveKeysend {
+			payment_data,
+			payment_preimage,
+			payment_metadata,
+			incoming_cltv_expiry: outgoing_cltv_value,
+			custom_tlvs,
+		}
+	} else if let Some(data) = payment_data {
+		PendingHTLCRouting::Receive {
+			payment_data: data,
+			payment_metadata,
+			incoming_cltv_expiry: outgoing_cltv_value,
+			phantom_shared_secret,
+			custom_tlvs,
+		}
+	} else {
+		return Err(InboundOnionErr {
+			err_code: 0x4000|0x2000|3,
+			err_data: Vec::new(),
+			msg: "We require payment_secrets",
+		});
+	};
+	Ok(PendingHTLCInfo {
+		routing,
+		payment_hash,
+		incoming_shared_secret: shared_secret,
+		incoming_amt_msat: Some(amt_msat),
+		outgoing_amt_msat: onion_amt_msat,
+		outgoing_cltv_value,
+		skimmed_fee_msat: counterparty_skimmed_fee_msat,
+	})
+}
+
+/// Peel one layer off an incoming onion, returning PendingHTLCInfo (either Forward or Receive).
+/// This does all the relevant context-free checks that LDK requires for payment relay or
+/// acceptance. If the payment is to be received, and the amount matches the expected amount for
+/// a given invoice, this indicates the UpdateAddHTLC, once fully committed in the channel, will
+/// generate an Event::PaymentReceived.
+pub fn peel_payment_onion<
+	NS: Deref,
+	L: Deref,
+	T: secp256k1::Verification,
+>(
+	msg: &msgs::UpdateAddHTLC, node_signer: &NS, logger: &L, secp_ctx: &Secp256k1<T>,
+	cur_height: u32, accept_mpp_keysend: bool,
+) -> Result<PendingHTLCInfo, InboundOnionErr>
+where
+	NS::Target: NodeSigner,
+	L::Target: Logger,
+{
+	let (hop, shared_secret, next_packet_details_opt) = 
```

Comment:
nit: trailing whitespace

Reply:
ugh i'll be more careful with whitespace